pkg/kubelet/kubelet_pods.go

```diff
@@ -84,20 +86,34 @@ func (kl *Kubelet) getActivePods() []*v1.Pod {
 }
 
 // makeDevices determines the devices for the given container.
-// Experimental. For now, we hardcode /dev/nvidia0 no matter what the user asks for
-// (we only support one device per node).
-// TODO: add support for more than 1 GPU after #28216.
-func makeDevices(container *v1.Container) []kubecontainer.DeviceInfo {
-	nvidiaGPULimit := container.Resources.Limits.NvidiaGPU()
-	if nvidiaGPULimit.Value() != 0 {
-		return []kubecontainer.DeviceInfo{
-			{PathOnHost: "/dev/nvidia0", PathInContainer: "/dev/nvidia0", Permissions: "mrw"},
-			{PathOnHost: "/dev/nvidiactl", PathInContainer: "/dev/nvidiactl", Permissions: "mrw"},
-			{PathOnHost: "/dev/nvidia-uvm", PathInContainer: "/dev/nvidia-uvm", Permissions: "mrw"},
-		}
+// Experimental.
+func (kl *Kubelet) makeDevices(pod *v1.Pod, container *v1.Container) ([]kubecontainer.DeviceInfo, error) {
+	if container.Resources.Limits.NvidiaGPU().IsZero() {
+		return nil, nil
 	}
 
-	return nil
+	nvidiaGPUPaths, err := kl.gpuManager.AllocateGPU(pod, container)
+	if err != nil {
+		return nil, err
+	}
+	devices := []kubecontainer.DeviceInfo{
+		{
+			PathOnHost:      nvidia.NvidiaCtlDevice,
+			PathInContainer: nvidia.NvidiaCtlDevice,
+			Permissions:     "mrw",
+		},
+		{
+			PathOnHost:      nvidia.NvidiaUVMDevice,
+			PathInContainer: nvidia.NvidiaUVMDevice,
+			Permissions:     "mrw",
+		},
+	}
+
+	for i, path := range nvidiaGPUPaths {
+		devices = append(devices, kubecontainer.DeviceInfo{PathOnHost: path, PathInContainer: "/dev/nvidia" + strconv.Itoa(i), Permissions: "mrw"})
+	}
+
```

Comment:
This is a little bit tricky and that's why generally you don't want to deal with the GPU runtime inside Kubernetes.

Basically, you need the exact mapping `major:minor` and `/dev/nvidia<minor>` between the host and the container. Otherwise, the CUDA driver will try to fix its character devices at startup and will fail if it can't (e.g. the user specified `--cap-drop=MKNOD` in docker or have user namespaces enabled).

You also want to check whether UVM devices are present on the node, they might not be there if UVM is not loaded or your driver doesn't support UVM.